packages/engine/Source/Scene/VoxelTraversal.js

```diff
@@ -443,34 +465,33 @@ function requestData(that, keyframeNode) {
           keyframeNode.metadata[i] = data;
           // State is received only when all metadata requests have been received
           keyframeNode.state = KeyframeNode.LoadState.RECEIVED;
+          that._primitive.tileLoad.raiseEvent();
         } else {
           keyframeNode.state = KeyframeNode.LoadState.FAILED;
           break;
         }
       }
     }
+    if (keyframeNode.state === KeyframeNode.LoadState.FAILED) {
+      that._primitive.tileFailed.raiseEvent();
+    }
   }
 
   function postRequestFailure() {
     that._simultaneousRequestCount--;
     keyframeNode.state = KeyframeNode.LoadState.FAILED;
+    that._primitive.tileFailed.raiseEvent();
   }
 
-  const { keyframe, spatialNode } = keyframeNode;
-  const promise = provider.requestData({
-    tileLevel: spatialNode.level,
-    tileX: spatialNode.x,
-    tileY: spatialNode.y,
-    tileZ: spatialNode.z,
-    keyframe: keyframe,
-  });
+  const promise = provider.requestData(requestOptions);
 
   if (defined(promise)) {
     that._simultaneousRequestCount++;
     keyframeNode.state = KeyframeNode.LoadState.RECEIVING;
     promise.then(postRequestSuccess).catch(postRequestFailure);
   } else {
     keyframeNode.state = KeyframeNode.LoadState.FAILED;
+    that._primitive.tileFailed.raiseEvent();
```

Comment:
This may be a question beyond the scope of this PR, but when a provider returns `undefined` instead of a promise, is that actually a failure state? Typically when `undefined` is returned from a function like this, it signals that the request couldn't be scheduled this frame and will be tried again next frame.

CC @jjhembd

Reply:
For both `Cesium3DTilesVoxelProvider` and the procedural providers in the Sandcastles, an `undefined` return happened when the tile didn't exist. I agree that this is not really expected behavior.

In https://github.com/CesiumGS/cesium/pull/12432 I changed the providers to return a rejected Promise if the tile doesn't exist.

Reply:
I added a few comments in https://github.com/CesiumGS/cesium/pull/12432 that mentioned `undefined` may still be a valid return value that needs to be accounted for. But, yes, it should never be a failure state.

Reply:
> In https://github.com/CesiumGS/cesium/pull/12432 I changed the providers to return a rejected Promise if the tile doesn't exist.

This change here as described above https://github.com/CesiumGS/cesium/blob/41dbe65ce3746a4bb80962698ee49b6af0eca6bd/packages/engine/Source/Scene/VoxelTraversal.js#L429

This simplifies the logic on whether to raise the tileLoad or tileFailed event
Something like this at the end of the `loadTileSuccess` would be all that is needed.
```
    if (keyframeNode.state === KeyframeNode.LoadState.PROCESSING) {
      that._primitive.tileLoad.raiseEvent();
    }
    that._primitive.tileFailed.raiseEvent();
```